nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -1502,6 +2115,59 @@ pub mod test {
                 }
             }
         }
+
+        // All commissioned disks should have debug and zone root datasets.
+        for (sled_id, disk_config) in &blueprint.blueprint_disks {
+            for disk in &disk_config.disks {
+                let zpool = ZpoolName::new_external(disk.pool_id);
+                let datasets = datasets_for_sled(&blueprint, *sled_id);
+
+                let dataset =
+                    find_dataset(&datasets, &zpool, DatasetKind::Debug);
+                assert_eq!(
+                    dataset.disposition,
+                    BlueprintDatasetDisposition::InService
+                );
+                let dataset =
+                    find_dataset(&datasets, &zpool, DatasetKind::ZoneRoot);
+                assert_eq!(
+                    dataset.disposition,
+                    BlueprintDatasetDisposition::InService
+                );
+            }
+        }
+        // All zones should have dataset records.
+        for (sled_id, zone_config) in
+            blueprint.all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)
+        {
+            match blueprint.sled_state.get(&sled_id) {
+                // Decommissioned sleds don't keep dataset state around
```

Comment:
Is this check necessary? It seems surprising we could get a decommissioned sled back from `all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)`

Reply:
Good point, removed

Reply:
Hrm, removing these lines is triggering a test failure, investigating

Reply:
This fails the `test_prune_decommissioned_sleds` test.

This test appears to do the following:

1. Update a blueprint by setting a sled's policy to `Expunged`
2. This triggers the sled to still appear in the blueprint, but with a desired state of `Decommissioned`.
3. Looking at the test logs, this decommissioned sleds still has zones which are returned from `blueprint.all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)`
4. When we call `verify_blueprint` - where this code exists - we panic because we cannot find datasets for this sled

I *think* steps 1-3 are true regardless of this PR - they're kinda independent of datasets? I just added this check to the `verify_blueprint` function because I *wanted* to observe datasets.

From what I can tell, the transfer of "expunged sleds should cause zones to be expunged" is the responsibility of the `Planner`, which is build on top of the `BlueprintBuilder` being tested here.

Reply:
(I added this check back in https://github.com/oxidecomputer/omicron/pull/6229/commits/26c578dd66333975c82e881796f245dee7027687)